src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -355,19 +668,573 @@ private CategoryDataset buildDataSet(StaplerRequest req) {
+                 * build.
+                 * As of now only default metric i.e. "mean" is enabled but other metrics can also be
+                 * enabled by including them in the drop down menu provided on Jenkins UI.
+                 */
+                switch (metricName) {
+                    case THRESHOLD: {
+                        float threshold = 0.002f;
+                        moreLengthyTests += calculateLengthyTestsByThreshold(threshold, caseResult);
+                        break;
+                    }
+                    case MAX: {
+                        moreLengthyTests += calculateLengthyTestsByMax(caseResult, allTests);
+                        break;
+                    }
+                    case PREV: {
+                        moreLengthyTests += calculateLengthyTestsByPrev(caseResult, allTests);
+                        break;
+                    }
+                    default: {
+                        float alpha = 0.5f;
+                        moreLengthyTests += calculateLengthyTestsByMean(alpha, caseResult, allTests);
+                    }
+                }
+                lengthyTestCount += moreLengthyTests;
+                caseName = caseResult.getName();
+                if (!generateToolTip || moreLengthyTests == 0)
+                    continue;
+                else if (toolTipString.length() + caseName.length() > maxToolTipLength) {
+                    generateToolTip = false;
+                    toolTipString += ",...";
+                    continue;
+                }
+                if (!toolTipString.equals(""))
+                    toolTipString += ", ";
+                toolTipString += caseName;
+            }
+            NumberOnlyBuildLabel label = new NumberOnlyBuildLabel(testResultAction.run);
+            dataset.add(lengthyTestCount, "Lengthy Tests", label);
+            /*
+             * Also being stored in {@link #lengthyToolTip} in order to generate tooltips on hovering mouse
+             * over the trend.
+             */
+            lengthyToolTip.put(label, toolTipString);
+            stack.pop();
+        }
+        LOGGER.log(Level.FINER, "total test trend count for {0}: {1}", new Object[]{run, count});
+        return dataset.build();
+    }
+
+
+    /**
+     * Utility method to count number of test flappers in the the build present at the head of buildHistory.
+     * @param buildHistory Queue to represent the past builds considered to check whether a testcase flapped in any
+     *                     of the past builds. If it flapped then it will be counted as flapper.
+     * @param testsHistory Map representing the builds in which a testcase flapped for each testcase by storing a mapping
+     *                     from testcase to the queue of builds in which it flapped.
+     */
+    private void shiftBuildHistoryUtil (ArrayDeque<Pair<AbstractTestResultAction<?>, HashSet<Integer>>> buildHistory,
```

Comment:
include window in the method name.
`captureBuildWindow` ?

Reply:
Refactored it to `shiftBuildWindow`